### PR TITLE
Use the least busy agent as destination

### PR DESCRIPTION
**Pull request by matelakat**
This change allows people to select which agent to use as a destination while moving routers. By default the choice was random. This patch changes the default to select always the least busy agent. The least busy agent has the smallest number of routers on it. If people want the random behaviour, they can select it with the --agent-selection-mode=random command line switch.

The new method first asks for the list of routers on each of the target agents, and saves this information in a cache. Each time when an agent is picked, the method is assuming that the move was successful, and maintains it's internal state accordingly. The cache of routers per agent will expire after 5 minutes, in which case neutron will be queried again.

### CODE REVIEW

**Comment by matelakat** [comment]
[neutron-ha-tool: pick the least busy agent](https://trello.com/c/gbJcBnbO/2-neutron-ha-tool-pick-the-least-busy-agent)

File: files/default/neutron-ha-tool.py
Comment (aplanas) [inline files/default/neutron-ha-tool.py]: nit: I would use choice here too

File: files/default/README.md
Comment (aplanas) [inline files/default/README.md]: no every one uses virtualenvwrapper, so I would suggest to use `virtualenv` command

File: files/default/neutron-ha-tool.py
Comment (aplanas) [inline files/default/neutron-ha-tool.py]: {agent['id']: agent for agent in agents}

File: files/default/neutron-ha-tool.py
Comment (aplanas) [inline files/default/neutron-ha-tool.py]: you can drop `.keys()`

File: files/default/neutron-ha-tool.py
Comment (aplanas) [inline files/default/neutron-ha-tool.py]: you can ignore this, but I would use `{}`, `[]` or `()` to initialize empty dicts, lists or tuples, the reason is that is marginally faster and more usual

File: files/default/neutron-ha-tool.py
Comment (aplanas) [inline files/default/neutron-ha-tool.py]: because your measure unit is seconds, I would use `.seconds` field from the timedelta

File: files/default/neutron-ha-tool.py
Comment (aplanas) [inline files/default/neutron-ha-tool.py]: as a safety net, check `if agent_id_number_of_routers`?

File: files/default/README.md
Comment (matelakat) [reply files/default/README.md]: I'm removing those steps, and just assuming the reader is capable of creating an environment.

File: files/default/neutron-ha-tool.py
Comment (matelakat) [reply files/default/neutron-ha-tool.py]: nice, added

File: files/default/neutron-ha-tool.py
Comment (matelakat) [reply files/default/neutron-ha-tool.py]: done

File: files/default/neutron-ha-tool.py
Comment (matelakat) [reply files/default/neutron-ha-tool.py]: true, done

File: files/default/neutron-ha-tool.py
Comment (matelakat) [reply files/default/neutron-ha-tool.py]: done

File: files/default/neutron-ha-tool.py
Comment (matelakat) [reply files/default/neutron-ha-tool.py]: what happens here is an `IndexError` is thrown. Basically the same would happen if we used the original implementation with random:

```python
>>> import random
>>> random.choice([])
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/lib64/python2.7/random.py", line 275, in choice
    return seq[int(self.random() * len(seq))]  # raises IndexError if seq is empty
IndexError: list index out of range
```

So I think let's leave it as it is?

Comment (aplanas) [reply files/default/neutron-ha-tool.py]: We do not want to die in neutron-ha-tool, this will cause a fail in pacemaker and a migration. I would log

Comment (matelakat) [reply files/default/neutron-ha-tool.py]: @aplanas the caller makes sure always that the array's length is more than one. Before we call `migrate_l3_routers_from_agent` we always check if `targets` is a non-null list. I still vote for leaving this as it is.

Comment (aplanas) [reply files/default/neutron-ha-tool.py]: ok for me

File: files/default/test-neutron-ha-tool.py
Comment (aplanas) [inline files/default/test-neutron-ha-tool.py]: Why the import is here? there will be not 'reimport' in every `setUp`

File: files/default/test-neutron-ha-tool.py
Comment (aplanas) [inline files/default/test-neutron-ha-tool.py]: this, maybe, is also not part of the setUp, can be in the global side

File: files/default/test-neutron-ha-tool.py
Comment (aplanas) [inline files/default/test-neutron-ha-tool.py]: same

**Review by aplanas — APPROVED** [review]
Basically OK from me. Except maybe the `import logger`

File: files/default/test-neutron-ha-tool.py
Comment (aspiers) [inline files/default/test-neutron-ha-tool.py]: That looks like a pretty unconventional way of doing the import.  Isn't it possible to do it with a normal `import`?

File: files/default/test-neutron-ha-tool.py
Comment (aspiers) [inline files/default/test-neutron-ha-tool.py:107]: Doesn't this assume that `MockNeutronClient` starts with no agents?  That sounds a bit brittle - better to assert that as part of the test.

File: files/default/test-neutron-ha-tool.py
Comment (aspiers) [inline files/default/test-neutron-ha-tool.py]: Not alive but `admin_state_up` is `True`?  That sounds strange to me but I'm probably just ignorant - in what circumstances would that happen?

File: files/default/test-neutron-ha-tool.py
Comment (aspiers) [inline files/default/test-neutron-ha-tool.py:114]: Same here, better to assert your mock's starting point.

File: files/default/test-neutron-ha-tool.py
Comment (aspiers) [inline files/default/test-neutron-ha-tool.py:130]: Same here.

File: files/default/neutron-ha-tool.py
Comment (aspiers) [reply files/default/neutron-ha-tool.py]: If so, probably `choose` is better than `choice` since it should be a verb.  But if you change it, would be better to stay consistent and rename the class to `RandomAgentChooser`.  Personally I slightly prefer `Picker` but there's not much difference.

File: files/default/README.md
Comment (aspiers) [reply files/default/README.md]: To do this properly, isn't `tox` the best approach, consistent with upstream?

**Review by aspiers — CHANGES_REQUESTED** [review]
Great work thanks, especially the nice refactoring of random choice into a strategy class as a separate commit before introducing a second strategy! That's exactly how to make a review task really easy and pleasant to do.  

However please don't add extra "address review comments" commits, as that makes the review process a lot harder.  Fortunately [I wrote a blog post about this](http://blog.adamspiers.org/2015/03/24/why-and-how-to-correctly-amend-github-pull-requests/) last year, so please read that.

File: files/default/neutron-ha-tool.py
Comment (aplanas) [reply files/default/neutron-ha-tool.py]: I suggested `choice` because of the python API. I guess that `pick` is ok, even to my eyes is too similar to pickle

File: files/default/test-neutron-ha-tool.py
Comment (matelakat) [reply files/default/test-neutron-ha-tool.py]: problem is that the filename contains dashes. I looked for other solutions, and will switch to using the `importlib` module, that might read a bit better. The good solution would be to rename the file so it does not contain dashes.

Comment (aplanas) [reply files/default/test-neutron-ha-tool.py]: I do not know a better way neither

File: files/default/test-neutron-ha-tool.py
Comment (matelakat) [reply files/default/test-neutron-ha-tool.py:107]: Yes, no agents, therefore none of them are dead, so it should not do anything. Let me amend the test, add one live agent, so it makes more sense what we are testing here. But in reality the test highlights that the tool will return 0 if we have no agents at all in the system. Whether that's a good behaviour or not - it's a different question.

File: files/default/test-neutron-ha-tool.py
Comment (matelakat) [reply files/default/test-neutron-ha-tool.py]: The criteria for a live agent is that both `alive` and `admin_state_up` are `True`. I think I better make the `admin_state_up` to be `False` as well here. I must admit, I'm not fully aware what values could pop up here. I set one of them `False`, so that the agent will be reported as dead.

Comment (matelakat) [reply files/default/test-neutron-ha-tool.py]: I looked at it - `alive` is reported by the agent periodically, so if the server doesn't get an `alive` message from the agent within a period of time, it will set `alive` to false. It's kind of a valid error scenario, but for now I think it's better if these initial tests reflect the happy usecases. We can add further tests to see what's the tool's response in the edge cases.

File: files/default/test-requirements.txt
Comment (Itxaka) [inline files/default/test-requirements.txt:1]: cant see where coverage or flake8 is being used here :)

Comment (aplanas) [reply files/default/test-requirements.txt:1]: Was documented in the readme.md, and is a good step before travis integration. IMHO it is not hurting.

Comment (Itxaka) [reply files/default/test-requirements.txt:1]: Dumb people(like me) dont read the readmes! Duh!

Disregard comment, sorry!

Comment (matelakat) [reply files/default/test-requirements.txt:1]: I think there is still a valid point, so I'll add the flake8 to the readme as well.

**Review by rhafer — APPROVED** [review]
Nice work! Don't have much to add to the existing  reviews. However I slightly wonder if putting the README.md and unittests into the files/default/ directory is a good choice. Or if its about time to cleanup this repository/branch and remove all the chef-specific bits from it. This is of course nothing that needs to happen as part of this PR, merely some food for thought.

We also should make sure to not merge this one before https://github.com/SUSE-Cloud/cookbook-openstack-network/pull/7 (I guess it's better to add the ```do not merge yet``` label here for now).

**Comment by matelakat** [comment]
@aspiers commits have been re-worked a bit, and I made the tests easier to read by adding some helper functions.

**Comment by Itxaka** [comment]
small thing @matelakat , we should add *.pyc files to the .gitignore files as running the tests will generate them and we dont want someone checking them in git by mistake :)

**Review by Itxaka — APPROVED** [review]

File: files/default/neutron-ha-tool.py
Comment (vuntz) [inline files/default/neutron-ha-tool.py]: Not sure if it really matters, but strictly speaking, I guess having a `cache_life.days > 0` check wouldn't hurt.

File: files/default/neutron-ha-tool.py
Comment (vuntz) [inline files/default/neutron-ha-tool.py:962]: Shouldn't we check that the agent is still alive? (it's actually not specific to this class, was also true for the random case) I can imagine the case where an agent dies, neutron-ha-tool is called to migrate things synchronously, but because it's synchronous, it takes some time and a second agent dies... Or is it purely theoretical?

File: files/default/neutron-ha-tool.py
Comment (matelakat) [reply files/default/neutron-ha-tool.py]: Right, good spot. Look at this:
```python
>>> import datetime
>>> now =datetime.datetime.now()
>>> tomorrow = now + datetime.timedelta(days=1)
>>> (tomorrow - now).seconds
0
>>> (tomorrow - now).total_seconds()
86400.0
```

I guess I'll use `total_seconds` as I did initially <- @aplanas

Comment (aplanas) [reply files/default/neutron-ha-tool.py]: Oh I see! I was wrong here and `.total_seconds` is correct.

File: files/default/neutron-ha-tool.py
Comment (matelakat) [reply files/default/neutron-ha-tool.py:962]: I think you have a valid point there @vuntz , but I technically see that as a separate feature. What I'm thinking about is adding an agent checker that will check the agents according to some method. That would mean that the agent pickers would need to support on-the fly agent removal.

Let's see what happens now: I believe when we poll the port status, we will time out for that router and for all the routers that were targeted to the same agent. So our actual response is timing out horribly, and on the next run the dead agent is detected. This is a bit far from optimal. I think we would need to address it, as we'll have higher timeouts already in the system due to this synchronous nature of router movement.

A fail-fast approach: when polling the ports, also poll the agent for it's liveness. When we find that id died, we raise. Next run will take care of it.

So another approach might be: When we poll the ports, we should also poll the agent status. If we find that the agent has died in the meanwhile, that triggers:
 - Put the actual router back to the queue
 - Re-list the live agents

Let me create a separate card for this item.

**Comment by matelakat** [comment]
@aspiers Let me know if I have the green light on this.

File: files/default/neutron-ha-tool.py
Comment (vuntz) [inline files/default/neutron-ha-tool.py:114]: It sounds to me like the output of this may not be the three lines you have here -- couldn't it happen that the first two lines are merged, which will result in some not so nice text?

**Review by vuntz — COMMENTED** [review]
You may consider this as +1; just one comment where I'm not 100% sure that I'm correct.

File: files/default/neutron-ha-tool.py
Comment (aspiers) [inline files/default/neutron-ha-tool.py:117]: flake8 is separate to unit testing so should have been in a separate commit.  I was initially confused by trying to figure out how these whitespace changes were related to unit testing, before realising that they are not.

File: files/default/neutron-ha-tool.py
Comment (aspiers) [inline files/default/neutron-ha-tool.py:783]: What's this `noqa` comment for?

File: files/default/neutron-ha-tool.py
Comment (aspiers) [inline files/default/neutron-ha-tool.py:795]: Same question here.

File: files/default/neutron-ha-tool.py
Comment (aspiers) [inline files/default/neutron-ha-tool.py]: Hmm.  This isn't really much more than a glorified global variable which just happens to have been wrapped inside a class, and in general use of globals should be treated with caution.  It lets you avoid passing the configuration around via method calls, but that results in the usage scope not being clearly defined, and other disadvantages associated with globals.  I won't give a -1 since it probably doesn't cause any real harm here, but I'd recommend thinking twice before using this design pattern in the future.  http://wiki.c2.com/?GlobalVariablesConsideredHarmful and http://wiki.c2.com/?GlobalVariablesAreBad are interesting reading ...

File: files/default/neutron-ha-tool.py
Comment (aspiers) [inline files/default/neutron-ha-tool.py:24]: GitHub doesn't let me review the commit message directly, but it should say "its" not "it's" ;-)  "it's" always means "it is".

File: files/default/neutron-ha-tool.py
Comment (aspiers) [inline files/default/neutron-ha-tool.py]: Is it good practice for a class to "steal" methods from another class like this?  I don't know, but I'm dubious it is, because it changes the context in which the method is executed.  If the method relies on internal state (which it probably doesn't in this case) then couldn't that be evaluated in the wrong context?  I would have thought that

    def now(self):
        return datetime.datetime.now()

would be safer and less voodoo, but I'd be happy to hear that I'm wrong.

File: files/default/neutron-ha-tool.py
Comment (aspiers) [reply files/default/neutron-ha-tool.py:962]: Very good discussion here, thanks guys :)

**Review by aspiers — APPROVED** [review]
Great work!  Still a few opportunities to polish this further if you are feeling diligent, but it's probably already good enough to merge.  I'd appreciate responses to my questions either way, though.

File: files/default/neutron-ha-tool.py
Comment (aspiers) [reply files/default/neutron-ha-tool.py:114]: @vuntz Nope, the `argparse` module automatically does line-wrapping by default, although I admit that this is quite unclear from the docs:

https://docs.python.org/2/library/argparse.html#formatter-class

File: files/default/neutron-ha-tool.py
Comment (matelakat) [reply files/default/neutron-ha-tool.py:783]: To silence the flake8 violation on this line: `E712 comparison to True should be 'if cond is True:' or 'if cond:'`. I wasn't brave enough to fix the violation itself, as I see that as a separate change, so for now to silence the violation, I added the `#noqa`, see https://pypi.python.org/pypi/flake8 .

File: files/default/neutron-ha-tool.py
Comment (matelakat) [reply files/default/neutron-ha-tool.py]: The intention of having now being injectable was because I wanted to have the ability to replace `now` with a mock should we need. Looking at the tests, this was never used, so it's a candidate for YAGNI now. At a later change this has been removed also. I might move that change here, so we don't do the add-and-remove cycle.

File: files/default/neutron-ha-tool.py
Comment (matelakat) [reply files/default/neutron-ha-tool.py:24]: > GitHub doesn't let me review the commit message directly, but it should say "its" not "it's" ;-) "it's" always means "it is".

Thanks, will amend it.

File: files/default/neutron-ha-tool.py
Comment (matelakat) [reply files/default/neutron-ha-tool.py:117]: You are right, they should have been in separate commits - Let me split them - although to be fair, the commit message said that it contains both flake8 and unit testing.

**Comment by vuntz** [comment]
@matelakat not sure if there are still changes pending here? If not, I guess you can merge since you got 4 approvals :-)

**Comment by matelakat** [comment]
@vuntz - @aspiers had some comments. For some reason they are only visible when you review the change commit-by commit.

**Pull request merged**
